native/dbdsqr.go

```diff
@@ -0,0 +1,493 @@
+					}
+					if ncc > 0 {
+						impl.Dlasr(blas.Left, lapack.Variable, lapack.Backward, m-ll, ncc, work[0:], work[n-1:], c[ll*ldc:], ldc)
+					}
+					if math.Abs(e[ll]) <= thresh {
+						e[ll] = 0
+					}
+				}
+			} else {
+				// Use nonzero shift.
+				if idir == 1 {
+					// Chase bulge from top to bottom. Save cosines and sines for
+					// later singular vector updates.
+					f := (math.Abs(d[ll]) - shift) * (math.Copysign(1, d[ll]) + shift/d[ll])
+					g := e[ll]
+					var cos1, sin1 float64
+					for i := ll; i < m-1; i++ {
+						cosr, sinr, r := impl.Dlartg(f, g)
+						if i > ll {
+							e[i-1] = r
+						}
+						f = cosr*d[i] + sinr*e[i]
+						e[i] = cosr*e[i] - sinr*d[i]
+						g = sinr * d[i+1]
+						d[i+1] *= cosr
+						cos1, sin1, r = impl.Dlartg(f, g)
+						d[i] = r
+						f = cos1*e[i] + sin1*d[i+1]
+						d[i+1] = cos1*d[i+1] - sin1*e[i]
+						if i < m-2 {
+							g = sin1 * e[i+1]
+							e[i+1] = cos1 * e[i+1]
+						}
+						work[i-ll] = cosr
+						work[i-ll+nm1] = sinr
+						work[i-ll+nm12] = cos1
+						work[i-ll+nm13] = sin1
+					}
+					e[m-2] = f
+					if ncvt > 0 {
+						impl.Dlasr(blas.Left, lapack.Variable, lapack.Forward, m-ll, ncvt, work[0:], work[n-1:], vt[ll*ldvt:], ldvt)
+					}
+					if nru > 0 {
+						impl.Dlasr(blas.Right, lapack.Variable, lapack.Forward, nru, m-ll, work[nm12:], work[nm13:], u[ll:], ldu)
+					}
+					if ncc > 0 {
+						impl.Dlasr(blas.Left, lapack.Variable, lapack.Forward, m-ll, ncc, work[nm12:], work[nm13:], c[ll*ldc:], ldc)
+					}
+					if math.Abs(e[m-2]) <= thresh {
+						e[m-2] = 0
+					}
+				} else {
+					// Chase bulge from top to bottom. Save cosines and sines for
+					// later singular vector updates.
+					f := (math.Abs(d[m-1]) - shift) * (math.Copysign(1, d[m-1]) + shift/d[m-1])
+					g := e[m-2]
+					for i := m - 1; i >= ll+1; i-- {
+						cosr, sinr, r := impl.Dlartg(f, g)
+						if i < m-1 {
+							e[i] = r
+						}
+						f = cosr*d[i] + sinr*e[i-1]
+						e[i-1] = cosr*e[i-1] - sinr*d[i]
+						g = sinr * d[i-1]
+						d[i-1] *= cosr
+						cos1, sin1, r := impl.Dlartg(f, g)
+						d[i] = r
+						f = cos1*e[i-1] + sin1*d[i-1]
+						d[i-1] = cos1*d[i-1] - sin1*e[i-1]
+						if i > ll+1 {
+							g = sin1 * e[i-2]
+							e[i-2] *= cos1
+						}
+						work[i-ll-1] = cosr
+						work[i-ll+nm1-1] = -sinr
+						work[i-ll+nm12-1] = cos1
+						work[i-ll+nm13-1] = -sin1
+					}
+					e[ll] = f
+					if math.Abs(e[ll]) <= thresh {
+						e[ll] = 0
+					}
+					if ncvt > 0 {
+						impl.Dlasr(blas.Left, lapack.Variable, lapack.Backward, m-ll, ncvt, work[nm12:], work[nm13:], vt[ll*ldvt:], ldvt)
+					}
+					if nru > 0 {
+						impl.Dlasr(blas.Right, lapack.Variable, lapack.Backward, nru, m-ll, work[0:], work[n-1:], u[ll:], ldu)
+					}
+					if ncc > 0 {
+						impl.Dlasr(blas.Left, lapack.Variable, lapack.Backward, m-ll, ncc, work[0:], work[n-1:], c[ll*ldc:], ldc)
+					}
+				}
+			}
+		}
+		// All singular values converged, make them positive.
```

Comment:
This comment seems out of place. Move below the closing brace, presumably just above the following for loop - add a blank line before the comment.